src/rhsmlib/cloud/_base_provider.py

```diff
@@ -84,32 +84,41 @@ class attributes. Logic of gathering metadata/signature will be implemented
     # Time to live of in-memory cache for metadata and signature (value is in seconds)
     IN_MEMORY_CACHE_TTL = 10.0
 
+    # Timeout for connection with IMDS server. The value is in seconds. Default value 1.0 second
+    # should be enough, because IMDS server is usually in the same datacenter and delay should
+    # be in milliseconds
+    TIMEOUT = 1.0
+
     def __init__(self, hw_info: dict = None):
         """
         Initialize cloud provider
         :param hw_info: Dictionary with hardware information.
         """
         # In-memory cache of metadata
-        self._cached_metadata: str = None
+        self._cached_metadata: Union[str, None] = None
         # Time, when metadata was received. The value is in seconds (unix time)
-        self._cached_metadata_ctime: float = None
+        self._cached_metadata_ctime: Union[float, None] = None
         # In-memory cache of signature
-        self._cached_signature: str = None
+        self._cached_signature: Union[str, None] = None
         # Time, when signature was received. The value is in seconds (unix time)
-        self._cached_signature_ctime: float = None
+        self._cached_signature_ctime: Union[float, None] = None
 
         # Dictionary with hardware information
         if hw_info is None:
             self.hw_info: dict = self.collect_hw_facts()
         else:
             self.hw_info: dict = hw_info
 
+        # HTTP Session
+        self._session = requests.Session()
+        self._session.hooks['response'].append(self._cb_debug_print_http_response)
```

Comment:
I'd say to set this only if `SUBMAN_DEBUG_PRINT_RESPONSE` is in `os.environ`, so we avoid adding overhead in normal (non-debug) runtime; this allows to drop the same check in `_cb_debug_print_http_response()`

Reply:
Yeah, it is not possible to alter environment variable in running process. Thus it make sense to add this hook only in situation, when the environment variable is set.